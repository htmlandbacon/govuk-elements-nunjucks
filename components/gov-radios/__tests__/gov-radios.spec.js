'use strict';

const nunjucks = require('nunjucks');
const cheerio = require('cheerio');

const yesNoRadios = [{id: 'yes', label: 'Yes', value: 'yes'}, {id: 'no', label: 'No', value: 'no'}];
const yesNoMaybeRadios = [{id: 'yes', label: 'Yes', value: 'yes'}, {id: 'no', label: 'No', value: 'no'}, {id: 'maybe', label: 'Maybe', value: 'maybe'}];
describe('gov-radios', () => {
  it('should set the ID as input-{name} if ID attr not supplied and generate radio buttons', () => {
    const name = 'selection';
    const output = nunjucks.render('./components/gov-radios/template.njk', {
      legend: 'Select a radio button',
      radioButtons: yesNoRadios,
      name
    });
    const $ = cheerio.load(output);
    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`input-${name}-yes`);
    expect(inputs.get(1).attribs.id).toBe(`input-${name}-no`);
    expect(output).toMatchSnapshot();
  });

  it('should use ID attribute value over generated input-{name} ID and generate radio buttons', () => {
    const id = 'my-id';
    const output = nunjucks.render('./components/gov-radios/template.njk', {
      legend: 'Select a radio button',
      radioButtons: yesNoRadios,
      name: 'selection',
      id
    });
    const $ = cheerio.load(output);
    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`${id}-yes`);
    expect(inputs.get(1).attribs.id).toBe(`${id}-no`);
    expect(output).toMatchSnapshot();
  });

  it('should add a form hint using the hint attribute', () => {
    const hint = 'Select one of the below';
    const output = nunjucks.render('./components/gov-radios/template.njk', {
      legend: 'Select a radio button',
      radioButtons: yesNoRadios,
      name: 'selection',
      id: 'my-id',
      hint
    });
    const $ = cheerio.load(output);
    const hintText = $('legend span.form-hint').text();
    expect(hintText).toBe(hint);

    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`my-id-yes`);
    expect(inputs.get(1).attribs.id).toBe(`my-id-no`);
    expect(output).toMatchSnapshot();
  });

  it('should set the radio button to checked if the value matches', () => {
    const value = 'no';
    const output = nunjucks.render('./components/gov-radios/template.njk', {
      legend: 'Select a radio button',
      radioButtons: yesNoRadios,
      name: 'selection',
      id: 'my-id',
      value
    });
    const $ = cheerio.load(output);
    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    const firstRadio = inputs.get(0).attribs;
    const secondRadio = inputs.get(1).attribs;
    expect(firstRadio.id).toBe(`my-id-yes`);
    expect(firstRadio.checked).toBe(undefined);
    expect(secondRadio.id).toBe(`my-id-no`);
    expect(secondRadio.checked).toBe('');
    expect(output).toMatchSnapshot();
  });

  it('should add error message and classes when passed an error object', () => {
    const error = 'Select an option';
    const output = nunjucks.render('./components/gov-radios/template.njk', {
      legend: 'Select a radio button',
      radioButtons: yesNoRadios,
      name: 'selection',
      id: 'my-id',
      error
    });
    const $ = cheerio.load(output);

    const inputs = $('input');
    expect(inputs).toHaveLength(2);
    expect(inputs.get(0).attribs.id).toBe(`my-id-yes`);
    expect(inputs.get(1).attribs.id).toBe(`my-id-no`);

    const formGroupClasses = $('.form-group').attr('class');
    const errorMsg = $('legend span.error-message').text();

    expect(formGroupClasses).toBe('form-group form-group-error');
    expect(errorMsg).toBe(error);
    expect(output).toMatchSnapshot();
  });
  describe('radio button options', () => {
    it('should add data fields when supplied', () => {
      const radioButton = [{id: 'id',
        value: 'value-id',
        label: 'label-id',
        dataTarget: 'data-target',
        dataJourneyClick: 'data-target-click'}];

      const output = nunjucks.render('./components/gov-radios/template.njk', {
        legend: 'Select a radio button',
        radioButtons: radioButton,
        id: 'my-id',
        name: 'selection'
      });
      const $ = cheerio.load(output);

      const firstDivContainer = $('div.multiple-choice');
      expect(firstDivContainer.get(0).attribs['data-target']).toBe('data-target');
      expect(firstDivContainer.get(0).attribs['data-journey-click']).toBe('data-target-click');

      const firstLabel = $('label');
      const labelCheck = expect.stringContaining('label-id');
      expect(firstLabel.text()).toEqual(labelCheck);
      const firstRadio = $('input').get(0).attribs;
      expect(firstRadio.id).toBe('my-id-id');
      expect(firstRadio.name).toBe('selection');
      expect(firstRadio.value).toBe('value-id');
      expect(output).toMatchSnapshot();
    });
  });
  describe('class options', () => {
    it('should add classes to legend span', () => {
      const classes = {legend: 'test-legend-class'};
      const output = nunjucks.render('./components/gov-radios/template.njk', {
        legend: 'Select a radio button',
        radioButtons: yesNoRadios,
        name: 'selection',
        id: 'my-id',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('.form-group legend span').attr('class');

      expect(labelClasses).toBe('form-label test-legend-class');
      expect(output).toMatchSnapshot();
    });
    it('should add inline class to form to fieldset when two radio buttons are supplied', () => {
      const classes = {legend: 'test-legend-class'};
      const output = nunjucks.render('./components/gov-radios/template.njk', {
        legend: 'Select a radio button',
        radioButtons: yesNoRadios,
        name: 'selection',
        id: 'my-id',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('fieldset').attr('class');

      expect(labelClasses).toBe('inline');
      expect(output).toMatchSnapshot();
    });
    it('should have no classes on fieldset when two or more radio buttons are supplied', () => {
      const classes = {legend: 'test-legend-class'};
      const output = nunjucks.render('./components/gov-radios/template.njk', {
        legend: 'Select a radio button',
        radioButtons: yesNoMaybeRadios,
        name: 'selection',
        id: 'my-id',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('fieldset').attr('class');

      expect(labelClasses).toBe(undefined);
      expect(output).toMatchSnapshot();
    });
  });
});

