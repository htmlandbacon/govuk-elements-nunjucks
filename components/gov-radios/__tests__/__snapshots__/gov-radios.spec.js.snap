// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gov-radios class options should add classes to legend span 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label test-legend-class">Select a radio button</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-yes" type="radio" name="selection" value="yes" />
      <label for="my-id-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-no" type="radio" name="selection" value="no" />
      <label for="my-id-no">No</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios class options should add inline class to form to fieldset when two radio buttons are supplied 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label test-legend-class">Select a radio button</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-yes" type="radio" name="selection" value="yes" />
      <label for="my-id-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-no" type="radio" name="selection" value="no" />
      <label for="my-id-no">No</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios class options should have no classes on fieldset when two or more radio buttons are supplied 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset>
    <legend>
      <span class="form-label test-legend-class">Select a radio button</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-yes" type="radio" name="selection" value="yes" />
      <label for="my-id-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-no" type="radio" name="selection" value="no" />
      <label for="my-id-no">No</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-maybe" type="radio" name="selection" value="maybe" />
      <label for="my-id-maybe">Maybe</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios radio button options should add data fields when supplied 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label">Select a radio button</span>
    </legend>
    <div class="multiple-choice" data-target="data-target" data-journey-click="data-target-click">
      <input id="my-id-id" type="radio" name="selection" value="value-id" />
      <label for="my-id-id">label-id</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios should add a form hint using the hint attribute 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label">Select a radio button</span>
      <span class="form-hint">Select one of the below</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-yes" type="radio" name="selection" value="yes" />
      <label for="my-id-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-no" type="radio" name="selection" value="no" />
      <label for="my-id-no">No</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios should add error message and classes when passed an error object 1`] = `
<div class="form-group error" id="my-id-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label">Select a radio button</span>
      <span class="error-message">Select an option</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-yes" type="radio" name="selection" value="yes" />
      <label for="my-id-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-no" type="radio" name="selection" value="no" />
      <label for="my-id-no">No</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios should set the ID as input-{name} if ID attr not supplied and generate radio buttons 1`] = `
<div class="form-group" id="input-selection-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label">Select a radio button</span>
    </legend>
    <div class="multiple-choice">
      <input id="input-selection-yes" type="radio" name="selection" value="yes"
      />
      <label for="input-selection-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="input-selection-no" type="radio" name="selection" value="no"
      />
      <label for="input-selection-no">No</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios should set the radio button to checked if the value matches 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label">Select a radio button</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-yes" type="radio" name="selection" value="yes" />
      <label for="my-id-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-no" type="radio" name="selection" value="no" checked
      />
      <label for="my-id-no">No</label>
    </div>
  </fieldset>
</div>
`;

exports[`gov-radios should use ID attribute value over generated input-{name} ID and generate radio buttons 1`] = `
<div class="form-group" id="my-id-form">
  <fieldset class="inline">
    <legend>
      <span class="form-label">Select a radio button</span>
    </legend>
    <div class="multiple-choice">
      <input id="my-id-yes" type="radio" name="selection" value="yes" />
      <label for="my-id-yes">Yes</label>
    </div>
    <div class="multiple-choice">
      <input id="my-id-no" type="radio" name="selection" value="no" />
      <label for="my-id-no">No</label>
    </div>
  </fieldset>
</div>
`;
